File: historyscan-back/src/main/java/io/sch/historyscan/domain/contexts/analysis/network/CodebaseNetworkAnalysis.java
package io.sch.historyscan.domain.contexts.analysis.network;

import io.sch.historyscan.domain.contexts.analysis.clocrevisions.CodebaseClocRevisions;
import io.sch.historyscan.domain.contexts.analysis.common.Analyze;
import io.sch.historyscan.domain.contexts.analysis.common.CodeBaseToAnalyze;
import io.sch.historyscan.domain.contexts.analysis.history.CodeBaseHistory;
import io.sch.historyscan.domain.error.HistoryScanFunctionalException;
import io.sch.historyscan.domain.hexagonalarchitecture.DDDService;

@DDDService
public class CodebaseNetworkAnalysis implements Analyze<CodebaseRevisionsNetwork> {

    private final Analyze<CodebaseClocRevisions> clocRevisionsAnalysis;
    private final Analyze<CodeBaseHistory> historyAnalysis;

    public CodebaseNetworkAnalysis(
            Analyze<CodebaseClocRevisions> clocRevisionsAnalysis,
            Analyze<CodeBaseHistory> historyAnalysis) {
        this.clocRevisionsAnalysis = clocRevisionsAnalysis;
        this.historyAnalysis = historyAnalysis;
    }

    @Override
    public CodebaseRevisionsNetwork from(CodeBaseToAnalyze codeBase) throws HistoryScanFunctionalException {
        var history = historyAnalysis.from(codeBase);
        var clocRevisions = clocRevisionsAnalysis.from(codeBase);
        return new CodebaseRevisionsNetwork(history, clocRevisions)
                .calculateNetworkFromHistoryAndRevisions();
    }
}

File: historyscan-back/src/main/kotlin/io/sch/historyscan/domain/contexts/analysis/network/CodebaseNetworkAnalysis.kt
package io.sch.historyscan.domain.contexts.analysis.network

import io.sch.historyscan.domain.contexts.analysis.clocrevisions.CodebaseClocRevisions
import io.sch.historyscan.domain.contexts.analysis.common.*
import io.sch.historyscan.domain.contexts.analysis.history.CodeBaseHistory
import io.sch.historyscan.domain.error.HistoryScanFunctionalException
import io.sch.historyscan.domain.hexagonalarchitecture.DDDService

@DDDService
class CodebaseNetworkAnalysis(
    private val clocRevisionsAnalysis: Analyze<CodebaseClocRevisions>,
    private val historyAnalysis: Analyze<CodeBaseHistory>
) : Analyze<CodebaseRevisionsNetwork?> {
    @Throws(HistoryScanFunctionalException::class)
    override fun from(codeBaseToAnalyze: CodeBaseToAnalyze): CodebaseRevisionsNetwork {
        val history = historyAnalysis.from(codeBaseToAnalyze)
        val clocRevisions = clocRevisionsAnalysis.from(codeBaseToAnalyze)
        return CodebaseRevisionsNetwork(history, clocRevisions)
            .calculateNetworkFromHistoryAndRevisions(codeBaseToAnalyze)
    }
}
